server/task_executor_pitr_restore.go

```diff
@@ -0,0 +1,162 @@
+		return true, nil, err
+	}
+
+	// Insert a pending baseline migration record into the migration_history table.
+	// The whole PITR process can be regarded as a data change followed by a schema sync.
+	migrationHistoryID, err := util.BeginMigration(ctx, executor, migrationInfo, prevSchemaBuf.String(), "", migrationInfo.Database)
+	if err != nil {
+		return true, nil, err
+	}
+	startedNs := time.Now().UnixNano()
+
+	var updatedSchemaBuf bytes.Buffer
+	if _, err := executor.Dump(ctx, migrationInfo.Database, &updatedSchemaBuf, true /*schemaOnly*/); err != nil {
+		return true, nil, util.FormatError(err)
+	}
+	updatedSchema := updatedSchemaBuf.String()
+
+	if err := exec.doPITRRestore(ctx, task, server.store, driver); err != nil {
+		return true, nil, err
+	}
+
+	// TODO(dragonly): Fix this using shared context between tasks.
+	if err := util.EndMigration(ctx, exec.l, executor, startedNs, migrationHistoryID, updatedSchema, db.BytebaseDatabase, true /*isDone*/); err != nil {
+		exec.l.Error("failed to update migration history record",
+			zap.Int64("migration_id", migrationHistoryID),
+			zap.Error(err),
+		)
+	}
+
+	exec.l.Info("created PITR database", zap.String("target database", migrationInfo.Database))
+
+	return true, &api.TaskRunResultPayload{
+		Detail:      fmt.Sprintf("Created PITR database for target database %q", migrationInfo.Database),
+		MigrationID: migrationHistoryID,
+		Version:     migrationInfo.Version,
+	}, nil
+}
+
+func (exec *PITRRestoreTaskExecutor) doPITRRestore(ctx context.Context, task *api.Task, store *store.Store, driver db.Driver) error {
+	instance := task.Instance
+	database := task.Database
+
+	issue, err := store.GetIssueByPipelineID(ctx, task.PipelineID)
+	if err != nil || issue == nil {
+		exec.l.Error("failed to get issue by PipelineID",
+			zap.Int("PipelineID", task.PipelineID),
+			zap.Error(err))
+		return fmt.Errorf("failed to get issue by PipelineID[%d], error[%w]", task.PipelineID, err)
+	}
+
+	mysqlDriver, ok := driver.(*pluginmysql.Driver)
+	if !ok {
+		exec.l.Error("failed to cast driver to mysql.Driver", zap.Stack("stack"))
+		return fmt.Errorf("[internal] cast driver to mysql.Driver failed")
+	}
+
+	mysqlRestore := restoremysql.New(mysqlDriver)
+	config := pluginmysql.BinlogInfo{}
+	// TODO(dragonly): Search and put the file io of the logical backup file here.
+	// Currently, let's just use the empty backup dump as a placeholder.
+	var buf bytes.Buffer
+	buf.WriteString("-- This is a fake backup dump")
+
+	exec.l.Debug("Start creating and restoring PITR database",
+		zap.String("instance", instance.Name),
+		zap.String("database", database.Name),
+	)
+
+	// RestorePITR will create the pitr database.
+	// Since it's ephemeral and will be renamed to the original database soon, we will reuse the original
+	// database's migration history, and append a new BASELINE migration.
+	if err := mysqlRestore.RestorePITR(ctx, bufio.NewScanner(&buf), config, database.Name, issue.CreatedTs); err != nil {
```

Comment:
We may want to rename the timestamp argument of RestorePITR to something else to differentiate it from targetTime.

Reply:
duplicated.